<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [15]:
import pandas as pd
import psycopg2

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [16]:
connection = psycopg2.connect(
    dbname='',
    user='',
    host='',
    password='',
    port=
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [17]:
# текст запроса
query_3_1 = f'''select count(*)
                from vacancies
             '''

In [18]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipykernel_118565/1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [19]:
# текст запроса
query_3_2 = f'''select count(*)
                from employers
             '''

In [20]:
# результат запроса
df2 = pd.read_sql_query(query_3_2, connection)
df2

/tmp/ipykernel_118565/142250957.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df2 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [21]:
# текст запроса
query_3_3 = f'''select count(*)
                from areas
             '''

In [22]:
# результат запроса
df3 = pd.read_sql_query(query_3_3, connection)
df3

/tmp/ipykernel_118565/1177936900.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df3 = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [23]:
# текст запроса
query_3_4 = f'''select count(*)
                from industries
             '''

In [24]:
# результат запроса
df4 = pd.read_sql_query(query_3_4, connection)
df4

/tmp/ipykernel_118565/1823014487.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4 = pd.read_sql_query(query_3_4, connection)


count
0    294

***

In [ ]:
# выводы по предварительному анализу данных
На мой взгляд количество вакансий слишком маленькое для такого количества работодателей.
Необходим более детельный анализ.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [362]:
# текст запроса
# Соединяем таблицы vacancies и areas, группируем по полю name таблицы areas
query_4_1 = f'''select count(v.id) cnt,
                       a.name area
                from vacancies v
                left join areas a on a.id = v.area_id
                group by a.name
                order by 1 desc
             '''

In [363]:
# результат запроса
df5 = pd.read_sql_query(query_4_1, connection)
df5

/tmp/ipykernel_118565/1863695066.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5 = pd.read_sql_query(query_4_1, connection)


cnt             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [31]:
# текст запроса
query_4_2 = f'''select count(*)
                from vacancies
                where salary_from is not null or salary_to is not null
             '''

In [32]:
# результат запроса
df6 = pd.read_sql_query(query_4_2, connection)
df6

/tmp/ipykernel_118565/1688172426.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6 = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [366]:
# текст запроса
# Выбираем из таблицы vacancies не пустые поля salary_from и salary_to, считаем среднее  и округляем до целого
query_4_3 = f'''select round(avg(salary_from), 0) avg_from,
                       round(avg(salary_to), 0) avg_to
                from vacancies
                where salary_from is not null or salary_to is not null
             '''

In [367]:
# результат запроса
df7 = pd.read_sql_query(query_4_3, connection)
df7

/tmp/ipykernel_118565/282124316.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df7 = pd.read_sql_query(query_4_3, connection)


avg_from    avg_to
0   71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [368]:
# текст запроса
# Группируем таблицу vacancies по  полям schedule и employment, считаем количество вакансий
query_4_4 = f'''select count(v.id) cnt,
                      v.schedule shed,
                      v.employment emp
                from vacancies v
                group by v.schedule,v.employment
                order by 1 desc
             '''

In [369]:
# результат запроса
df8 = pd.read_sql_query(query_4_4, connection)
df8

/tmp/ipykernel_118565/2222883958.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df8 = pd.read_sql_query(query_4_4, connection)


cnt              shed                  emp
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [370]:
# текст запроса
query_4_5 = f'''select count(v.id) cnt,
                      v.experience exp
                      
                from vacancies v
                group by v.experience
                order by 1
             '''

In [371]:
# результат запроса
df9 = pd.read_sql_query(query_4_5, connection)
df9

/tmp/ipykernel_118565/1145373401.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df9 = pd.read_sql_query(query_4_5, connection)


cnt                 exp
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

In [ ]:
# выводы по детальному анализу вакансий
После проведенного анализа вакансий можно сделать следующие выводы:
1. Количество вакансий коррелирует с количеством жителей населенного пункта.
2. Достаточно большое количество вакансий имеют не заполненные поля ЗП.
3. Вилка по ЗП достаточно большая, скорее всего имеет место связь размера ЗП и населенного пункта, где размещена вакансия.
4. Основная часть вакансий предполагает полную занятость с различными вариантами рабочего графика.
5. Основная часть вакансий предполагает опыт соискателей от 1 до 6 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [372]:
# текст запроса
# Соединяем таблицы vacancies  и employers группируем по полю name таблицы employers, считаем количество вакансий
query_5_1 = f'''select count(v.id) cnt,
                       e.name emp_name
                from vacancies v
                left join employers e on e.id = v.employer_id
                group by e.name
                order by 1 desc
                limit 5
             '''

In [373]:
# результат запроса
df10 = pd.read_sql_query(query_5_1, connection)
df10

/tmp/ipykernel_118565/415274739.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df10 = pd.read_sql_query(query_5_1, connection)


cnt       emp_name
0  1933         Яндекс
1   491     Ростелеком
2   444       Тинькофф
3   428           СБЕР
4   331  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [374]:
# текст запроса
# Соединяем таблицы areas и vacancies, employers группируем по полю name таблицы areas, считаем количество вакансий
query_5_2 = f'''select count(v.id) cnt_vac,
                       count(e.id) cnt_emp,
                       a.name area
                from areas a
                left join employers e on a.id = e.area
                left join vacancies v on a.id = v.area_id
                group by a.name
                order by 1, 2 desc
                
             '''

In [375]:
# результат запроса
df11 = pd.read_sql_query(query_5_2, connection)
df11

/tmp/ipykernel_118565/2599469216.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df11 = pd.read_sql_query(query_5_2, connection)


cnt_vac   cnt_emp                area
0            0       410              Россия
1            0       207           Казахстан
2            0        75  Московская область
3            0        19  Краснодарский край
4            0        18  Ростовская область
...        ...       ...                 ...
1357   1149438   1149438         Новосибирск
1358   1364132   1364132              Алматы
1359   2354880   2354880               Минск
1360   6320667   6320667     Санкт-Петербург
1361  31272712  31272712              Москва

[1362 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [376]:
# текст запроса
# Соединяем таблицы vacancies и employers группируем по полю name таблицы employers, считаем уникальное количество регионов
query_5_3 = f'''select count(distinct v.area_id) cnt,
                       e.name emp_name
                from vacancies v
                left join employers e on e.id = v.employer_id
                group by e.name
                order by 1 desc
               
             '''

In [377]:
# результат запроса
df12 = pd.read_sql_query(query_5_3, connection)
df12

/tmp/ipykernel_118565/3418963348.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df12 = pd.read_sql_query(query_5_3, connection)


cnt                emp_name
0      181                  Яндекс
1      152              Ростелеком
2      116              Спецремонт
3       88  Поляков Денис Иванович
4       71               ООО ЕФИН 
...    ...                     ...
14761    1                  UniSol
14762    1            UNISTORY LLC
14763    1                   UNIT6
14764    1     United Distribution
14765    1              UNITEL LLC

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [378]:
# текст запроса
query_5_4 = f'''select count(e.id) cnt
                       
                from employers e
                left join employers_industries i on e.id = i.employer_id
                where i.employer_id is null
             '''

In [379]:
# результат запроса
df13 = pd.read_sql_query(query_5_4, connection)
df13

/tmp/ipykernel_118565/3773637443.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df13 = pd.read_sql_query(query_5_4, connection)


cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [380]:
# текст запроса
# Соединяем таблицы employers_industries и employers группируем по полю name таблицы employers, считаем количество сфер деятельности
# выводим работодалей где сфер деятельности 4
query_5_5 = f'''select count(i.industry_id) cnt,
                       e.name
                from employers_industries i
                left join employers e on e.id = i.employer_id
                group by e.name
                having count(i.industry_id)=4
                order by 2
             '''

In [381]:
# результат запроса
df14 = pd.read_sql_query(query_5_5, connection)
df14

/tmp/ipykernel_118565/2950930045.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df14 = pd.read_sql_query(query_5_5, connection)


cnt             name
0       4     101 Интернет
1       4         21vek.by
2       4             2ГИС
3       4               2К
4       4      4 пикселя +
...   ...              ...
1131    4            ЮРИОН
1132    4  ЮТИП Технологии
1133    4           ЯКласс
1134    4            ЯрНео
1135    4    Яшин&Партнёры

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [382]:
# текст запроса
query_5_6 = f'''select count(distinct i.employer_id) cnt
                       
                from employers_industries i
                left join industries e on e.id = i.industry_id
                where e.name = 'Разработка программного обеспечения'
             '''

In [383]:
# результат запроса
df15 = pd.read_sql_query(query_5_6, connection)
df15

/tmp/ipykernel_118565/2033316979.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df15 = pd.read_sql_query(query_5_6, connection)


cnt
0  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [384]:
# код для получения списка городов-милионников
# Для получения содержания страницы используем модуль wikipedia
# Для получения данных использует страницу на русском языке
# Читаем таблицы из обьекта html в датафрейм df из татфрейма получаем в список значения колонки Город
import pandas as pd
import wikipedia as wp
wp.set_lang("ru")
html = wp.page("Города-миллионеры России").html()
df = pd.read_html(html)[1]
city = df['Город'].tolist()
print (city)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


/tmp/ipykernel_118565/3624900399.py:9: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  df = pd.read_html(html)[1]


In [385]:
# текст запроса
# Программно создаем текст запроса в зависимости от количества элементов списка городов молионников
# В цикле создаются однотипные запросы с различным значением отбора по полю  name таблицы areas
# Также в цикле создаем условие для запроса выводящего итоговые значения вакансий
# Запросы обьединяем с помощью union
q_all = ""
cond = ""
for c in city:
    q = ""
    q = f'''select  a.name, count(v.id) cnt from vacancies v left join employers e on e.id = v.employer_id left join areas a on a.id = v.area_id where  e.name = 'Яндекс' and a.name = '{c}' group by a.name union ''' 
    q_all = q_all + q
    if cond == "" :
        cond = cond +  f''' a.name = '{c}' '''
    else:
        cond = cond + 'or ' + f''' a.name = '{c}' '''

q2 = f'''select  'total', count(v.id) cnt from vacancies v left join employers e on e.id = v.employer_id left join areas a on a.id = v.area_id where ({cond}) and e.name = 'Яндекс'   order by 2'''

q_all = q_all + q2

query_5_7 = f'''{q_all}
             '''

In [386]:
# результат запроса
df16 = pd.read_sql_query(query_5_7, connection)
df16

/tmp/ipykernel_118565/4126436065.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df16 = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4    Ростов-на-Дону   25
5            Казань   25
6             Пермь   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            total  485

***

In [ ]:
# выводы по анализу работодателей
После анализа работодателей, можно сделать следующие выводы:
1. Наибольшее количество вакансий размещено компанией Яндекс, при этом вакансии достаточно равномерно распределены по регионам.
2. Имеет место ситуация когда, при размещении вакансии регион Москва включает в себя Московскую область.
3. Крупные компании такие как Яндекс, Сбер, Газпромнефть имеют много направлений деятельности и соответственно размещают вакансии по различным направлениям.
4. В базе достаточно много компаний занимаются разработкой ПО. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [360]:
# текст запроса
query_6_1 = f'''select count(*) cnt
                from vacancies
                where lower(name) LIKE '%data%'
                      OR lower(name) LIKE '%данн%'
             '''

In [361]:
# результат запроса
df17 = pd.read_sql_query(query_6_1, connection)
df17

/tmp/ipykernel_118565/3982197767.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df17 = pd.read_sql_query(query_6_1, connection)


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [387]:
# текст запроса
# С помощью LIKE отбираем вакансии удовлетворяющие условиям
query_6_2 = f'''select count(*) cnt
                from vacancies
                where (lower(name) LIKE '%data scientist%'
                      OR lower(name) LIKE '%data science%'
                      OR lower(name) LIKE '%исследователь данных%'
                      OR (name LIKE '%ML%' and name NOT LIKE '%HTML%')
                      OR lower(name) LIKE '%machine learning%'
                      OR lower(name) LIKE '%машинн%обучен%')  
                      AND (lower(name) LIKE '%junior%'  
                          OR lower(experience) LIKE '%нет опыта%'
                          OR lower(schedule) LIKE '%стажировка%')
             '''

In [388]:
# результат запроса
df18 = pd.read_sql_query(query_6_2, connection)
df18

/tmp/ipykernel_118565/1277669926.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df18 = pd.read_sql_query(query_6_2, connection)


cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [389]:
# текст запроса
query_6_3 = f'''select count(*) cnt
                from vacancies
                where ((lower(name) LIKE '%data scientist%'
                      OR lower(name) LIKE '%data science%'
                      OR lower(name) LIKE '%исследователь данных%'
                      OR (name LIKE '%ML%' and name NOT LIKE '%HTML%')
                      OR lower(name) LIKE '%machine learning%'
                      OR lower(name) LIKE '%машинн%обучен%') 
                      AND (key_skills LIKE '%SQL%' OR key_skills LIKE '%postgres%'))
             '''

In [390]:
# результат запроса
df19 = pd.read_sql_query(query_6_3, connection)
df19

/tmp/ipykernel_118565/1247231042.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df19 = pd.read_sql_query(query_6_3, connection)


cnt
0  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [391]:
# текст запроса
query_6_4 = f'''select count(*) cnt
                from vacancies
                where (lower(name) LIKE '%data scientist%'
                      OR lower(name) LIKE '%data science%'
                      OR lower(name) LIKE '%исследователь данных%'
                      OR (name LIKE '%ML%' and name NOT LIKE '%HTML%')
                      OR lower(name) LIKE '%machine learning%'
                      OR lower(name) LIKE '%машинн%обучен%') 
                      AND (lower(key_skills) LIKE '%python%') 
             '''

In [392]:
# результат запроса
df20 = pd.read_sql_query(query_6_4, connection)
df20

/tmp/ipykernel_118565/3009142906.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df20 = pd.read_sql_query(query_6_4, connection)


cnt
0  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [393]:
# текст запроса
# Условия отбора однотипны для всех вакансий  DS, методика подсчета ключевых навыков взята из подсказки
query_6_5 = f'''select round(avg((LENGTH(key_skills) - LENGTH(replace(key_skills,CHR(9),''))) + 1), 2) cnt
                from vacancies
                where (key_skills IS NOT NULL)
                       AND (lower(name) LIKE '%data scientist%'  
                       OR lower(name) LIKE '%data science%' 
                       OR lower(name) LIKE '%исследователь данных%' 
                       OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')  
                       OR lower(name) LIKE '%machine learning%' 
                       OR lower(name) LIKE '%машинн%обучен%')  
            '''

In [394]:
# результат запроса
df21 = pd.read_sql_query(query_6_5, connection)
df21

/tmp/ipykernel_118565/3069874189.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df21 = pd.read_sql_query(query_6_5, connection)


cnt
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [395]:
# текст запроса
# условия отбора те же, с учетом что одно из полей _from,salary_to не NULL
# считаем ЗП с помощью coalesce, если одно из полей NULL подставляем второе 
query_6_6 = f'''select experience,
                       round(avg((coalesce(salary_from,salary_to) + coalesce(salary_to, salary_from))/2), 0) zp
                from vacancies
                where (salary_from IS NOT NULL OR salary_to IS NOT NULL) 
                      AND (lower(name) LIKE '%data scientist%'  
                      OR lower(name) LIKE '%data science%' 
                      OR lower(name) LIKE '%исследователь данных%' 
                      OR (name LIKE '%ML%' and name NOT LIKE '%HTML%')  
                      OR lower(name) LIKE '%machine learning%' 
                      OR lower(name) LIKE '%машинн%обучен%') 
                group by experience
            '''

In [396]:
# результат запроса
df22 = pd.read_sql_query(query_6_6, connection)
df22

/tmp/ipykernel_118565/710439576.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df22 = pd.read_sql_query(query_6_6, connection)


experience        zp
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

In [ ]:
# выводы по предметному анализу
На основании проведенных исследований можно сделать следующие выводы:
1. Вакансий в сфере DS достаточно малое количество (1771 если конкретизировать запрос то 480), хотелось бы посмотреть на более свежие данные,
чтобы оценить динамику (рост, падение).
2. Рынок труда требует специалистов с опытом работы, что видно как по количеству вакансий, так и по сдредней ЗП.
3. Работодатели размещающие вакансии в сфере DS крупные, многопрофильные компании. Это косвенно свидетельствует о широкой применимости DS.  



# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

In [397]:
# текст запроса
# С помощью LIKE отбираем вакансии удовлетворяющие условиям
query_6_7 = f'''select count(*) cnt
                from vacancies
                where (lower(name) LIKE '%data scientist%'
                      OR lower(name) LIKE '%data science%'
                      OR lower(name) LIKE '%исследователь данных%'
                      OR (name LIKE '%ML%' and name NOT LIKE '%HTML%')
                      OR lower(name) LIKE '%machine learning%'
                      OR lower(name) LIKE '%машинн%обучен%')  
                 '''

In [398]:
df23 = pd.read_sql_query(query_6_7, connection)
df23

/tmp/ipykernel_118565/1851832912.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df23 = pd.read_sql_query(query_6_7, connection)


cnt
0  480

Общее количество вакансий DS равно 480  из которых в 351 упомянут Python и 201 упомянуты SQL или postgres.
На основании этого можно сделать вывод о важности этих навыков в сфере DS.

In [401]:
# текст запроса
# условия отбора те же, с учетом что одно из полей _from,salary_to не NULL
# считаем ЗП с помощью coalesce, если одно из полей NULL подставляем второе 
query_6_8 = f'''select v.experience,
                       a.name,
                       round(avg((coalesce(v.salary_from,v.salary_to) + coalesce(v.salary_to, v.salary_from))/2), 0) zp
                from vacancies v
                left join areas a on a.id = v.area_id
                where (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL) 
                      AND (lower(v.name) LIKE '%data scientist%'  
                      OR lower(v.name) LIKE '%data science%' 
                      OR lower(v.name) LIKE '%исследователь данных%' 
                      OR (v.name LIKE '%ML%' and v.name NOT LIKE '%HTML%')  
                      OR lower(v.name) LIKE '%machine learning%' 
                      OR lower(v.name) LIKE '%машинн%обучен%') 
                group by v.experience,a.name
                order by 3
            '''

In [402]:
df24 = pd.read_sql_query(query_6_8, connection)
df24

/tmp/ipykernel_118565/2144289808.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df24 = pd.read_sql_query(query_6_8, connection)


experience             name        zp
0            Нет опыта     Екатеринбург   37000.0
1   От 1 года до 3 лет           Казань   50000.0
2   От 1 года до 3 лет   Ростов-на-Дону   50000.0
3   От 1 года до 3 лет          Барнаул   60000.0
4        От 3 до 6 лет           Алматы   67126.0
5            Нет опыта      Владивосток   70000.0
6            Нет опыта           Москва   72500.0
7   От 1 года до 3 лет            Пермь   75000.0
8            Нет опыта      Новосибирск   80000.0
9            Нет опыта  Санкт-Петербург   87667.0
10  От 1 года до 3 лет              Уфа   90000.0
11  От 1 года до 3 лет      Владивосток  100000.0
12  От 1 года до 3 лет           Рязань  105000.0
13  От 1 года до 3 лет           Алматы  122047.0
14  От 1 года до 3 лет          Королев  130000.0
15       От 3 до 6 лет           Рязань  152500.0
16  От 1 года до 3 лет     Екатеринбург  167500.0
17  От 1 года до 3 лет           Москва  169179.0
18  От 1 года до 3 лет  Санкт-Петербург  175173.0
19       От 3 до 6 лет         Белгород  200000.0
20       От 3 до 6 лет      Новосибирск  209500.0
21       От 3 до 6 лет       Черногория  233794.0
22       От 3 до 6 лет           Сербия  233794.0
23       От 3 до 6 лет           Турция  233794.0
24       От 3 до 6 лет           Москва  263750.0
25       От 3 до 6 лет          Армения  268863.0
26       От 3 до 6 лет  Санкт-Петербург  300000.0
27       От 3 до 6 лет             Кипр  300000.0

Данный запрос показывает взаимосвязь средней ЗП не только от опыта соискателя, но и региона.

Как видно из полученных результатов, специалисты в области DS требуются крупным компаниям ведущим многопрофильную деятельность.
Предпочтение отдается работникам с опытом работы, однако и для специалистов уровня Junior присутствует некоторое количество вакансий.
Крупные компании ищут работников во всех городах милионниках России, однако уровень ЗП при одних и тех же требованиях разный.
По моему мнению, при дальнейшем проникновении технологий ИИ в различные сферы деятельности, потребность в специалистах в области DS будет только расти. Количество вакансий будет увеличиваться, кроме того размер компании будет менее крупным, чем в данный момент.  